File: src/main/java/com/pedigree/services/TagService.java
package com.pedigree.services;

import com.pedigree.model.Family;
import com.pedigree.model.Individual;
import com.pedigree.model.Tag;
import com.pedigree.storage.ProjectRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class TagService {
    private final ProjectRepository.ProjectData data;

    public TagService(ProjectRepository.ProjectData data) {
        this.data = Objects.requireNonNull(data, "data");
    }

    public void assignTagToIndividual(String individualId, Tag tag) {
        Individual individual = findIndividual(individualId).orElseThrow();
        if (individual.getTags().stream().noneMatch(t -> t.getId().equals(tag.getId()))) {
            individual.getTags().add(tag);
        }
        if (data.tags.stream().noneMatch(t -> t.getId().equals(tag.getId()))) {
            data.tags.add(tag);
        }
    }

    public void assignTagToFamily(String familyId, Tag tag) {
        Family family = findFamily(familyId).orElseThrow();
        if (family.getTags().stream().noneMatch(t -> t.getId().equals(tag.getId()))) {
            family.getTags().add(tag);
        }
        if (data.tags.stream().noneMatch(t -> t.getId().equals(tag.getId()))) {
            data.tags.add(tag);
        }
    }

    public void removeTagFromIndividual(String individualId, String tagId) {
        findIndividual(individualId).ifPresent(i -> i.getTags().removeIf(t -> t.getId().equals(tagId)));
    }

    public void removeTagFromFamily(String familyId, String tagId) {
        findFamily(familyId).ifPresent(f -> f.getTags().removeIf(t -> t.getId().equals(tagId)));
    }

    public List<Individual> filterIndividualsByTagName(String tagName) {
        return data.individuals.stream()
                .filter(i -> i.getTags().stream().anyMatch(t -> tagName.equalsIgnoreCase(t.getName())))
                .collect(Collectors.toList());
    }

    public List<Family> filterFamiliesByTagName(String tagName) {
        return data.families.stream()
                .filter(f -> f.getTags().stream().anyMatch(t -> tagName.equalsIgnoreCase(t.getName())))
                .collect(Collectors.toList());
    }

    public List<Tag> getAllTags() { return new ArrayList<>(data.tags); }

    private Optional<Individual> findIndividual(String id) {
        return data.individuals.stream().filter(i -> i.getId().equals(id)).findFirst();
    }

    private Optional<Family> findFamily(String id) {
        return data.families.stream().filter(f -> f.getId().equals(id)).findFirst();
    }
}

File: src/main/kotlin/com/family/tree/services/TagService.kt
package com.family.tree.services

import com.family.tree.model.Family
import com.family.tree.model.Individual
import com.family.tree.model.Tag
import com.family.tree.storage.ProjectRepository
import java.util.*

class TagService(private val data: ProjectRepository.ProjectData) {
    init { Objects.requireNonNull(data, "data") }

    fun assignTagToIndividual(individualId: String, tag: Tag) {
        val individual: Individual = findIndividual(individualId).orElseThrow()
        if (individual.tags.none { it.id == tag.id }) {
            individual.tags.add(tag)
        }
        if (data.tags.none { it.id == tag.id }) {
            data.tags.add(tag)
        }
    }

    fun assignTagToFamily(familyId: String, tag: Tag) {
        val family: Family = findFamily(familyId).orElseThrow()
        if (family.tags.none { it.id == tag.id }) {
            family.tags.add(tag)
        }
        if (data.tags.none { it.id == tag.id }) {
            data.tags.add(tag)
        }
    }

    fun removeTagFromIndividual(individualId: String, tagId: String) {
        findIndividual(individualId).ifPresent { i -> i.tags.removeIf { it.id == tagId } }
    }

    fun removeTagFromFamily(familyId: String, tagId: String) {
        findFamily(familyId).ifPresent { f -> f.tags.removeIf { it.id == tagId } }
    }

    fun filterIndividualsByTagName(tagName: String): List<Individual> {
        return data.individuals.filter { i -> i.tags.any { t -> tagName.equals(t.name, ignoreCase = true) } }
    }

    fun filterFamiliesByTagName(tagName: String): List<Family> {
        return data.families.filter { f -> f.tags.any { t -> tagName.equals(t.name, ignoreCase = true) } }
    }

    fun allTags(): List<Tag> = ArrayList(data.tags)

    private fun findIndividual(id: String) = data.individuals.stream().filter { it.id == id }.findFirst()
    private fun findFamily(id: String) = data.families.stream().filter { it.id == id }.findFirst()
}
